### PR TITLE
Support `:log`, `:log2`, `:log10` scales

### PR DESCRIPTION
**Pull request by t-bltg**
Fix https://github.com/JuliaPlots/UnicodePlots.jl/issues/32.

~~First naive tentative, not well tested (performance ?, correctness ?), looks good at first glance though.~~

**Example**
```julia
using UnicodePlots

show_nl(p) = (show(p); print('\n'^3))

main() = begin
  x = collect(1:100)
  y = collect(1:100)
  kw = (canvas=DotCanvas,)

  for func ∈ (scatterplot, lineplot)
    show_nl(func(x, y; xlabel="linear", ylabel="linear", title="$func | linear", kw...))

    show_nl(func(log10.(x), y; xlabel="log10", ylabel="linear", title="$func | semilog-x (expected)", kw...))
    show_nl(func(x, y; xlabel="log10", ylabel="linear", xscale=:log10, title="$func | semilog-x", kw...))

    show_nl(func(x, log10.(y); xlabel="linear", ylabel="log10", title="$func | semilog-y (expected)", kw...))
    show_nl(func(x, y; xlabel="linear", ylabel="log10", yscale=:log10, title="$func | semilog-y", kw...))

    show_nl(func(log10.(x), log10.(y); xlabel="log10", ylabel="log10", title="$func | loglog (expected)", kw...))
    show_nl(func(x, y; xlabel="log10", ylabel="log10", xscale=:log10, yscale=:log10, title="$func | loglog", kw...))
  end
  return
end

main()
```

**Output**
```julia
                        scatterplot | linear
              ┌────────────────────────────────────────┐ 
          100 │                                     .:'│ 
              │                                  ..:'  │ 
              │                                .:'     │ 
              │                             .:''       │ 
              │                          ..:'          │ 
              │                        .:'             │ 
              │                     .:''               │ 
   linear     │                  ..:'                  │ 
              │                .:'                     │ 
              │             .:''                       │ 
              │          ..:'                          │ 
              │        .:'                             │ 
              │     .:''                               │ 
              │  ..:'                                  │ 
            0 │.:'                                     │ 
              └────────────────────────────────────────┘ 
              0                                      100
                               linear


                 scatterplot | semilog-x (expected)
              ┌────────────────────────────────────────┐ 
          100 │                                       :│ 
              │                                      .:│ 
              │                                      : │ 
              │                                     :' │ 
              │                                    .'  │ 
              │                                   .'   │ 
              │                                  .'    │ 
   linear     │                                 .'     │ 
              │                                :'      │ 
              │                              .:'       │ 
              │                            .:'         │ 
              │                          .:'           │ 
              │                      ..:''             │ 
              │                . ..:''                 │ 
            0 │.     .  .  '' '                        │ 
              └────────────────────────────────────────┘ 
              0                                        2
                                log10


                       scatterplot | semilog-x
              ┌────────────────────────────────────────┐ 
          100 │                                       :│ 
              │                                      .:│ 
              │                                      : │ 
              │                                     :' │ 
              │                                    .'  │ 
              │                                   .'   │ 
              │                                  .'    │ 
   linear     │                                 .'     │ 
              │                                :'      │ 
              │                              .:'       │ 
              │                            .:'         │ 
              │                          .:'           │ 
              │                      ..:''             │ 
              │                . ..:''                 │ 
            0 │.     .  .  '' '                        │ 
              └────────────────────────────────────────┘ 
              0                                        2
                                log10


              scatterplot | semilog-y (expected)
           ┌────────────────────────────────────────┐ 
         2 │                             .....:'''''│ 
           │                      ...:''''          │ 
           │                ..:'''                  │ 
           │            ..''                        │ 
           │        ..:'                            │ 
           │      .:'                               │ 
           │    .:'                                 │ 
   log10   │   .:                                   │ 
           │  .'                                    │ 
           │  :                                     │ 
           │ '                                      │ 
           │ '                                      │ 
           │.                                       │ 
           │                                        │ 
         0 │.                                       │ 
           └────────────────────────────────────────┘ 
           0                                      100
                            linear


                    scatterplot | semilog-y
           ┌────────────────────────────────────────┐ 
         2 │                             .....:'''''│ 
           │                      ...:''''          │ 
           │                ..:'''                  │ 
           │            ..''                        │ 
           │        ..:'                            │ 
           │      .:'                               │ 
           │    .:'                                 │ 
   log10   │   .:                                   │ 
           │  .'                                    │ 
           │  :                                     │ 
           │ '                                      │ 
           │ '                                      │ 
           │.                                       │ 
           │                                        │ 
         0 │.                                       │ 
           └────────────────────────────────────────┘ 
           0                                      100
                            linear


                scatterplot | loglog (expected)
           ┌────────────────────────────────────────┐ 
         2 │                                     .:'│ 
           │                                  ..''  │ 
           │                                .:'     │ 
           │                             .:'        │ 
           │                          ..''          │ 
           │                        .:'             │ 
           │                     .:'                │ 
   log10   │                   .:                   │ 
           │                . '                     │ 
           │             . '                        │ 
           │            '                           │ 
           │         '                              │ 
           │      .                                 │ 
           │                                        │ 
         0 │.                                       │ 
           └────────────────────────────────────────┘ 
           0                                        2
                             log10


                     scatterplot | loglog
           ┌────────────────────────────────────────┐ 
         2 │                                     .:'│ 
           │                                  ..''  │ 
           │                                .:'     │ 
           │                             .:'        │ 
           │                          ..''          │ 
           │                        .:'             │ 
           │                     .:'                │ 
   log10   │                   .:                   │ 
           │                . '                     │ 
           │             . '                        │ 
           │            '                           │ 
           │         '                              │ 
           │      .                                 │ 
           │                                        │ 
         0 │.                                       │ 
           └────────────────────────────────────────┘ 
           0                                        2
                             log10


                          lineplot | linear
              ┌────────────────────────────────────────┐ 
          100 │                                     .:'│ 
              │                                  ..:'  │ 
              │                                .:'     │ 
              │                             .:''       │ 
              │                          ..:'          │ 
              │                        .:'             │ 
              │                     .:''               │ 
   linear     │                  ..:'                  │ 
              │                .:'                     │ 
              │             .:''                       │ 
              │          ..:'                          │ 
              │        .:'                             │ 
              │     .:''                               │ 
              │  ..:'                                  │ 
            0 │.:'                                     │ 
              └────────────────────────────────────────┘ 
              0                                      100
                               linear


                   lineplot | semilog-x (expected)
              ┌────────────────────────────────────────┐ 
          100 │                                       :│ 
              │                                      .:│ 
              │                                      : │ 
              │                                     :' │ 
              │                                    .'  │ 
              │                                   .'   │ 
              │                                  .'    │ 
   linear     │                                 .'     │ 
              │                                :'      │ 
              │                              .:'       │ 
              │                            .:'         │ 
              │                          .:'           │ 
              │                      ..:''             │ 
              │                ....:''                 │ 
            0 │..........''''''                        │ 
              └────────────────────────────────────────┘ 
              0                                        2
                                log10


                        lineplot | semilog-x
              ┌────────────────────────────────────────┐ 
          100 │                                       :│ 
              │                                      .:│ 
              │                                      : │ 
              │                                     :' │ 
              │                                    .'  │ 
              │                                   .'   │ 
              │                                  .'    │ 
   linear     │                                 .'     │ 
              │                                :'      │ 
              │                              .:'       │ 
              │                            .:'         │ 
              │                          .:'           │ 
              │                      ..:''             │ 
              │                ....:''                 │ 
            0 │..........''''''                        │ 
              └────────────────────────────────────────┘ 
              0                                        2
                                log10


                lineplot | semilog-y (expected)
           ┌────────────────────────────────────────┐ 
         2 │                             .....:'''''│ 
           │                      ...:''''          │ 
           │                ..:'''                  │ 
           │            ..''                        │ 
           │        ..:'                            │ 
           │      .:'                               │ 
           │    .:'                                 │ 
   log10   │   .:                                   │ 
           │  .'                                    │ 
           │ .:                                     │ 
           │ :                                      │ 
           │ :                                      │ 
           │:                                       │ 
           │:                                       │ 
         0 │:                                       │ 
           └────────────────────────────────────────┘ 
           0                                      100
                            linear


                     lineplot | semilog-y
           ┌────────────────────────────────────────┐ 
         2 │                             .....:'''''│ 
           │                      ...:''''          │ 
           │                ..:'''                  │ 
           │            ..''                        │ 
           │        ..:'                            │ 
           │      .:'                               │ 
           │    .:'                                 │ 
   log10   │   .:                                   │ 
           │  .'                                    │ 
           │ .:                                     │ 
           │ :                                      │ 
           │ :                                      │ 
           │:                                       │ 
           │:                                       │ 
         0 │:                                       │ 
           └────────────────────────────────────────┘ 
           0                                      100
                            linear


                 lineplot | loglog (expected)
           ┌────────────────────────────────────────┐ 
         2 │                                     .:'│ 
           │                                  ..''  │ 
           │                                .:'     │ 
           │                             .:'        │ 
           │                          ..''          │ 
           │                        .:'             │ 
           │                     .:'                │ 
   log10   │                   .:                   │ 
           │                .''                     │ 
           │             .''                        │ 
           │           .''                          │ 
           │        .:'                             │ 
           │      .'                                │ 
           │   ..'                                  │ 
         0 │..'                                     │ 
           └────────────────────────────────────────┘ 
           0                                        2
                             log10


                       lineplot | loglog
           ┌────────────────────────────────────────┐ 
         2 │                                     .:'│ 
           │                                  ..''  │ 
           │                                .:'     │ 
           │                             .:'        │ 
           │                          ..''          │ 
           │                        .:'             │ 
           │                     .:'                │ 
   log10   │                   .:                   │ 
           │                .''                     │ 
           │             .''                        │ 
           │           .''                          │ 
           │        .:'                             │ 
           │      .'                                │ 
           │   ..'                                  │ 
         0 │..'                                     │ 
           └────────────────────────────────────────┘ 
           0                                        2
                             log10
```

### CODE REVIEW

File: src/canvas/asciicanvas.jl [hunk 1]
```diff
@@ -111,6 +111,8 @@ struct AsciiCanvas <: LookupCanvas
     origin_y::Float64
     width::Float64
     height::Float64
+    xscale::Symbol
+    yscale::Symbol
```
Comment (johnnychen94) [inline src/canvas/asciicanvas.jl:115]: I know this is Plots convention to use symbols so no objections, but just let me raise a note here for this design: from a computational perspective, any scalar function can be put here (say, `abs2` or more complicated coordinate transformations).

**Review by johnnychen94 — COMMENTED** [review]
Nothing suspicious to me but I really can't say much on this because I know very little about plotting 😢 

I know you're trying to fix as many old issues as possible, yet it would still be necessary to have tests covered. How about testing if `lineplot(x, y)` and `lineplot(exp.(x), exp.(y); xscale=:log, yscale=:log)`? (I'm not sure if there will be tiny differences, however)

**Comment by t-bltg** [comment]
> How about testing if `lineplot(x, y)` and `lineplot(exp.(x), exp.(y); xscale=:log, yscale=:log)`?

Thanks, I couldn't get ideas right now for the tests, this is a good one ! Will add.

File: src/canvas/asciicanvas.jl (hunk 1 above)
Comment (t-bltg) [reply src/canvas/asciicanvas.jl]: This is now fixed, and tested.

**Review by johnnychen94 — APPROVED** [review]

**Pull request merged**